lensless/utils/dataset.py

```diff
@@ -0,0 +1,446 @@
+
+    """
+
+    def __init__(
+        self,
+        psf,
+        dataset,
+        pre_transform=None,
+        dataset_is_CHW=False,
+        flip=False,
+        **kwargs,
+    ):
+        """
+        Parameters
+        ----------
+
+        psf : torch.Tensor
+            PSF to use for simulate. Should be a 4D tensor with shape [1, H, W, C]. Simulation of multi-depth data is not supported yet.
+        dataset : torch.utils.data.Dataset
+            Dataset to propagate. Should output images with shape [H, W, C] unless ``dataset_is_CHW`` is ``True`` (and therefore images have the dimension ordering of [C, H, W]).
+        pre_transform : PyTorch Transform or None, optional
+            Transform to apply to the images before simulation, by default ``None``.
+        dataset_is_CHW : bool, optional
+            If True, the input dataset is expected to output images with shape [C, H, W], by default ``False``.
+        flip : bool, optional
+                If True,images are flipped beffore the simulation, by default ``False``..
+        """
+
+        # we do the flipping before the simualtion
+        super(SimulatedDataset, self).__init__(flip=False, **kwargs)
+
+        assert isinstance(dataset, Dataset)
+        self.dataset = dataset
+        self.n_files = len(dataset)
+
+        self.dataset_is_CHW = dataset_is_CHW
+        self._pre_transform = pre_transform
+        self.flip_pre_sim = flip
+
+        psf = psf.squeeze().movedim(-1, 0)
+
+        # initialize simulator
+        self.sim = FarFieldSimulator(psf=psf, is_torch=True, **kwargs)
+
+    def get_image(self, index):
+        return self.dataset[index]
+
+    def _get_images_pair(self, index):
+        # load image
+        img, _ = self.get_image(index)
+        # convert to CHW for simulator
+        if not self.dataset_is_CHW:
+            img = img.movedim(-1, 0)
+        if self.flip_pre_sim:
+            img = torch.rot90(img, dims=(-1, -2))
+        if self._pre_transform is not None:
+            img = self._pre_transform(img)
+
+        lensless, lensed = self.sim.propagate(img, return_object_plane=True)
+
+        # convert back to HWC
+        lensless = lensless.moveaxis(-3, -1)
+        lensed = lensed.moveaxis(-3, -1)
+
+        return lensless, lensed
+
+    def __len__(self):
+        if self.indices is None:
+            return self.n_files
+        else:
+            return len([x for x in self.indices if x < self.n_files])
+
+
+class LenslessDataset(DualDataset):
+    """
+    Dataset consisting of lensless image captured from a screen and the corresponding image shown on screen (not necessarily captured with a lens).
+    It can be used with a PyTorch DataLoader to load a batch of lensless and corresponding lensed images.
+    Unless the setup is perfectly calibrated, one should expect to have to use ``transform_lensed`` to adjust the alignement and rotation.
+    """
```

Comment:
Add that user should check parameters of FarFieldSimulator: https://github.com/ebezzam/waveprop/blob/c07863aac87a8cd9f90ad43aa8428eb185c1595b/waveprop/simulation.py#L11

Reply:
RST syntax for link
```
`Link text <link URL>`_

```